### DISCUSSION TITLE
Env support for Vercel cron jobs

### DISCUSSION THREAD

**Discussion by therecluse26** [vercel/community#2158] [Idea]
Hey there, reading through your docs on the recent cron offering, and the only security solution offered would involve committing a secret to source control. While this would work, the idea of doing that goes against security best practices. I think that the simplest solution would be to allow the interpolation of env variables such as:

```
{
    "crons": [
      {
        "path": "/api/send-slack-notification?key=${CRON_SLACK_KEY}",
        "schedule": "*/10 * * * *"
      },
    ]
  }
```

Thoughts?

**Comment by alexbjorlig** [comment]
This is exactly what I'm looking for - but I guess still not supported?

**Reply by therecluse26** [reply]
As far as I'm aware not currently, but hopefully soon

**Comment by vvo** [comment]
Hey there, protecting Cron Jobs on Vercel is now possible using this method:

1. Create an environment variable named `CRON_SECRET` in your Vercel Project
2. When present, we will invoke your Cron Job with an authorization header this way: `Authorization: Bearer $CRON_SECRET`
3. You can use this authorization header to verify the Cron Job was invoked by Vercel

Let me know if that's working well. We will announce this more broadly soon, along with updating docs.

**Reply by therecluse26** [reply]
Hey @vvo , thanks for the response! Unfortunately however, I'm not seeing how this would solve the issue of committing secrets (such as a Slack key as in the example) to source control. Could you please clarify? 

Are you saying that all environment variables are now accessible from the cron job, or just specifically `CRON_SECRET`?

**Reply by vvo** [reply]
> Are you saying that all environment variables are now accessible from the cron job

A Cron Job being a regular route, you already have access to all env variables defined on your Vercel project.

> I'm not seeing how this would solve the issue of committing secrets

You do not have to commit `CRON_SECRET`, all you have to do is add `CRON_SECRET` to your Vercel project environment variables (https://vercel.com/docs/concepts/projects/environment-variables) and then when invoking your Cron Job, if `CRON_SECRET` is part of your project env variables, then we will send it as part of the Cron Job payload.

Hope it makes a little more sense, let me know.

**Reply by nicolassanmar** [reply]
Thanks for the solution @vvo. Is this going to be added to the docs soon?

**Reply by AA-85** [reply]
Thanks @vvo ! This method of limiting access to cron jobs is very handy and tested to be working well

**Reply by vvo** [reply]
> Is this going to be added to the docs soon?

Yep it will @nicolassanmar

**Reply by bencmbrook** [reply]
The [example here](https://vercel.com/docs/cron-jobs/manage-cron-jobs#securing-cron-jobs) has a timing attack vulnerability

```ts
import type { NextRequest } from 'next/server';
 
export function GET(request: NextRequest) {
  const authHeader = request.headers.get('authorization');
  if (authHeader !== `Bearer ${process.env.CRON_SECRET}`) { // <-- not a constant time string comparison
    return new Response('Unauthorized', {
      status: 401,
    });
  }
 
  return Response.json({ success: true });
}
```

- Node has [`timingSafeEqual`](https://nodejs.org/docs/latest-v18.x/api/crypto.html#crypto_crypto_timingsafeequal_a_b)
- WebCrypto [doesn't](https://github.com/w3c/webcrypto/issues/270), but Cloudflare [added a polyfill](https://developers.cloudflare.com/workers/runtime-apis/web-crypto/#timingsafeequal) to the Cloudflare Workers runtime - maybe Vercel can do the same?

**Comment by Naeem-gg** [comment]
Hi @vvo , does Vercel cron jobs only support GET requests or we can invoke with other HTTP methods also?

**Reply by vvo** [reply]
Hey @Naeem-gg we call these endpoints via GET requests only at the moment. What's your usecase for other methods? Thanks 👍

**Reply by Naeem-gg** [reply]
To optimize database storage, I am scheduling a daily cron job to delete old unwanted records. I believe that a GET request may not be the most prudent approach for this critical operation. What do you think? or do you suggest an alternative method?

**Reply by vvo** [reply]
@Naeem-gg As long as you implement cron job verification you should be good in term of security for this: https://vercel.com/docs/cron-jobs/manage-cron-jobs#securing-cron-jobs
